test/JSValueScopeTests.cs

```diff
@@ -0,0 +1,348 @@
+        });
+        Assert.Equal(JSValueScopeType.Module, JSValueScope.Current.ScopeType);
+
+        using JSValueScope callbackScope = new(JSValueScopeType.Callback);
+        Assert.Throws<JSInvalidScopeException>(() =>
+        {
+            using JSValueScope rootScope = new(JSValueScopeType.Root);
+        });
+        Assert.Equal(JSValueScopeType.Callback, JSValueScope.Current.ScopeType);
+
+        using JSValueScope handleScope = new(JSValueScopeType.Handle);
+        Assert.Throws<JSInvalidScopeException>(() =>
+        {
+            using JSValueScope rootScope = new(JSValueScopeType.Root);
+        });
+        Assert.Equal(JSValueScopeType.Handle, JSValueScope.Current.ScopeType);
+
+        using JSValueScope escapableScope = new(JSValueScopeType.Escapable);
+        Assert.Throws<JSInvalidScopeException>(() =>
+        {
+            using JSValueScope rootScope = new(JSValueScopeType.Root);
+        });
+        Assert.Equal(JSValueScopeType.Escapable, JSValueScope.Current.ScopeType);
+    }
+
+    [Fact]
+    public void InvalidModuleContextScopeNesting()
+    {
+        using JSValueScope moduleScope = TestScope(JSValueScopeType.Module);
+        using JSValueScope callbackScope = new(JSValueScopeType.Callback);
+        Assert.Throws<JSInvalidScopeException>(() =>
+        {
+            using JSValueScope nestedModuleScope = new(JSValueScopeType.Module);
+        });
+        Assert.Equal(JSValueScopeType.Callback, JSValueScope.Current.ScopeType);
+
+        using JSValueScope handleScope = new(JSValueScopeType.Handle);
+        Assert.Throws<JSInvalidScopeException>(() =>
+        {
+            using JSValueScope nestedModuleScope = new(JSValueScopeType.Module);
+        });
+        Assert.Equal(JSValueScopeType.Handle, JSValueScope.Current.ScopeType);
+
+        using JSValueScope escapableScope = new(JSValueScopeType.Escapable);
+        Assert.Throws<JSInvalidScopeException>(() =>
+        {
+            using JSValueScope nestedModuleScope = new(JSValueScopeType.Module);
+        });
+        Assert.Equal(JSValueScopeType.Escapable, JSValueScope.Current.ScopeType);
+    }
+
+    [Fact]
+    public void InvalidCallbackContextScopeNesting()
+    {
+        using JSValueScope rootScope = TestScope(JSValueScopeType.Root);
+
+        using JSValueScope handleScope = new(JSValueScopeType.Handle);
+        Assert.Throws<JSInvalidScopeException>(() =>
+        {
+            using JSValueScope callbackScope = new(JSValueScopeType.Callback);
+        });
+        Assert.Equal(JSValueScopeType.Handle, JSValueScope.Current.ScopeType);
+
+        using JSValueScope escapableScope = new(JSValueScopeType.Escapable);
+        Assert.Throws<JSInvalidScopeException>(() =>
+        {
+            using JSValueScope callbackScope = new(JSValueScopeType.Callback);
+        });
+        Assert.Equal(JSValueScopeType.Escapable, JSValueScope.Current.ScopeType);
+    }
+
+    [Fact]
+    public void AccessValueFromClosedScope()
```

Comment:
Should we have a test that accesses `Hendle` property from `JSValue`?

Another interesting scenario would be when we have two nested scopes.
One scope creates and object and another one creates a property key.
Then, after the inner scope is closed, we use the key to set object property.
Since the Handle is never checked, the code will access the deleted memory for the property key.

Reply:
There is a test for accessing a value from a disposed scope. It is checked.

Reply:
The test is not covering a case when a function uses two `JSValue`s from different scopes, where one of the is disposed (property key) and the other is not (the object). Since the check for `napi_env` will succeed, the code will use deleted property key `napi_value`.

Reply:
That is a good point about secondary arguments. I added that test case.